extensions/notebook/src/book/bookModel.ts

```diff
@@ -184,29 +187,32 @@ export class BookModel implements azdata.nb.NavigationProvider {
 	 */
 	private parseJupyterSections(section: any[]): IJupyterBookSection[] {
 		try {
-			return section.reduce((acc, val) => Array.isArray(val.sections) ? acc.concat(val).concat(this.parseJupyterSections(val.sections)) : acc.concat(val), []);
-		} catch (error) {
-			let err: string = loc.invalidTocFileError(error);
+			return section.reduce((acc, val) => Array.isArray(val.sections) ?
+				acc.concat(val).concat(this.parseJupyterSections(val.sections)) : acc.concat(val), []);
+		} catch (e) {
+			this._errorMessage = loc.invalidTocFileError();
```

Comment:
Not passing in the error object anymore?

Reply:
error object fails with 'section.reduce is not a function' as long as section doesn't have a value. And that's not a useful error for us to show the user so I got rid of the error object.

Reply:
You should try to keep the book name in the error message somehow. Why don't we have access to it anymore? We should know what book we're trying to open when parsing the TOC at some level.

Reply:
book title is read from the toc file, thiserror happens when reading toc fails

Reply:
But we at least have the book path right? That would be good information to have. Possibly even direct link to the toc file (or the directory if it doesn't exist)

Reply:
added the book path to the error.

Reply:
This error still doesn't have that. If the toc doesn't have any sections we should default to showing the file name at least.

Reply:
the error isn't handled here, I have added the path where this error is actually being surfaced.

Reply:
Then why are we setting the error message here? It would be nice to try and spend a few minutes cleaning up the code flow here so it isn't quite so garbled of a mess - although if that ends up being a huge task I guess it's fine as it is. But if we're not using this error message for anything then let's at least try to get rid of it since right now it looks like something that would be surfaced to the user.

Reply:
No, it isn't a garbles mess. SO here's the flow: we call readbooks in which there are several calls that can fail. We have a try in there and catch errors there and summarize them. We definitely need to know if parseJupyterSections failed or yaml load failed. Here if parseJupyterSections fails we're throwing the error up and handlind it in the method that called it which has the path and tells that it actually failed to read books from the path due invalid toc.

Reply:
That's not a great way of doing error handling. Setting error messages in a single instance variable just makes things more convoluted and increases the chance that a wrong error is displayed - especially when we do so much async work.

And as mentioned exposing the error message like that for testing is also problematic for the same reasons. There's much better ways to verify that the correct error was thrown (actually throwing errors and catching them, using mocks to verify certain methods were called, etc.).

Anyways if you don't want to spend any time fixing this now that's ok, but let's not make this kind of code style a habit for future work.

Reply:
I would definitely want to do it the right way. This is not something I added, I have been fixing what Lucy had left from before. If you can point me to an example test where we're using mocks to verify, I can make mends.

Reply:
Doing so would require refactoring the code to get rid of the whole setting a custom error message thing IMO - and at this point I'm not sure it's worth the time investment.

But for future work here's some examples :

The simplest is using assert.throws : https://github.com/microsoft/azuredatastudio/blob/master/src/sql/workbench/services/connection/test/browser/connectionManagementService.test.ts#L657

The other way uses the ApiWrapper that we use to call into the vscode api - in this case likely something like showErrorMessage for when an error occurs. I don't have a specific example of this on-hand but in general it would just be a matter of setting up the mock for that function like we do here : https://github.com/microsoft/azuredatastudio/blob/master/extensions/azurecore/src/test/azureResource/providers/database/databaseTreeDataProvider.test.ts#L106 and then using verify to ensure that it was called : https://github.com/microsoft/azuredatastudio/blob/master/extensions/azurecore/src/test/azureResource/resourceTreeNode.test.ts#L157

Let me know if you still have any questions!

Reply:
Thanks this helps, let me see if it's worth spending the time to refactor. Will follow this next time on.